src/box/alter.cc

```diff
@@ -4395,6 +4396,207 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
+	bool with_snap;
+};
+
+/** Mapping from tuple.opts to fields of gc_consumer_def. */
+const struct opt_def gc_consumer_def_opts_reg[] = {
+	OPT_DEF("with_snap", OPT_BOOL, struct gc_consumer_def, with_snap),
+	OPT_END,
+};
+
+/**
+ * Fill gc_consumer_def with opts from the MsgPack map.
+ * Argument map can be NULL - default options are set in this case.
+ */
+static int
+gc_consumer_def_opts_decode(struct gc_consumer_def *def, const char *map,
+			    struct region *region)
+{
+	def->with_snap = false;
+	if (map == NULL)
+		return 0;
+	return opts_decode(def, gc_consumer_def_opts_reg, &map, region);
+}
+
+/** Build gc_consumer definition from a _gc_consumers' tuple. */
+static struct gc_consumer_def *
+gc_consumer_def_new_from_tuple(struct tuple *tuple, struct region *region)
+{
+	struct gc_consumer_def *def =
+		xregion_alloc_object(region, typeof(*def));
+	memset(def, 0, sizeof(*def));
+	if (tuple_field_uuid(tuple, BOX_GC_CONSUMERS_FIELD_UUID, &def->uuid) != 0)
+		return NULL;
+	if (tt_uuid_is_nil(&def->uuid)) {
+		diag_set(ClientError, ER_INVALID_UUID, tt_uuid_str(&def->uuid));
+		return NULL;
+	}
+	const char *mp_vclock =
+		tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_VCLOCK,
+				      MP_MAP);
+	if (mp_vclock == NULL)
+		return NULL;
+	if (mp_decode_vclock_ignore0(&mp_vclock, &def->vclock) != 0) {
+		diag_set(ClientError, ER_INVALID_VCLOCK);
+		return NULL;
+	}
+	const char *opts = NULL;
+	if (tuple_field(tuple, BOX_GC_CONSUMERS_FIELD_OPTS) != NULL) {
+		opts = tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_OPTS,
+					     MP_MAP);
+	}
+	if (gc_consumer_def_opts_decode(def, opts, region) != 0)
+		return NULL;
+	return def;
+}
+
+/**
+ * Data passed to transactional triggers of replace in _gc_consumers.
+ */
+struct gc_consumers_txn_trigger_data {
+	/*
+	 * Replica UUID. Is used instead of replica object because it can be
+	 * unregistered before the trigger is fired.
+	 */
+	struct tt_uuid uuid;
+	/*
+	 * New GC consumer, will be unregistered on rollback.
+	 */
+	struct gc_consumer *new_gc;
+};
+
+static int
+gc_consumers_rollback_insert_update(struct trigger *trigger, void *)
+{
+	struct gc_consumers_txn_trigger_data *data =
+		(struct gc_consumers_txn_trigger_data *)trigger->data;
+	/* Consumer was only created on replace, so simply delete it here. */
+	gc_consumer_unregister(data->new_gc);
+	return 0;
+}
+
+static int
+gc_consumers_commit_insert_update(struct trigger *trigger, void *)
+{
+	struct gc_consumers_txn_trigger_data *data =
+		(struct gc_consumers_txn_trigger_data *)trigger->data;
+	struct replica *replica = replica_by_uuid(&data->uuid);
+	/* Set new GC only if the replica is still registered. */
+	if (replica != NULL) {
+		/*
+		 * Drop the old gc consumer if any. If it's the first one,
+		 * unref gc_delay since this replica is already pinning xlogs.
+		 */
+		if (replica->gc != NULL)
+			gc_consumer_unregister(replica->gc);
+		else
+			gc_delay_unref();
+		replica->gc = data->new_gc;
```

Comment:
Sometimes it goes back, so I would patch `gc_consumer_advance` to allow it to be advanced in both directions. The real problem here is that I want to create a consumer on replace (do not wait for commit) and delete consumer on commit so that in the case of rollback no xlogs would be deleted, and `gc_consumer_advance` cannot be used then. But I agree that it's costly, so to make it much cheaper, I pushed a separate commit that replaces the `calloc` in `gc_consumer_register` with `mempool_alloc`.

Reply:
It happens only during `relay_subscribe`, as far as I remember. There we'll indeed have to recreate a consumer. Let's see what others will say. I'm ok with the current version too